LXD stop hook wrapper

tomponline opened the pull request:
This requires support in `lxd-user` to add support for the `callhook` subcommand.

Related to https://github.com/canonical/lxd/issues/13331

Depends on https://github.com/canonical/lxd/pull/14253

tomponline committed:
lxd-stophook: Adds stop hook wrapper for the lxd command

This command is intended to replace the `lxd` command in the snap so that
existing running containers that are stopping will call this script as part of
their stop hooks.

This will then "route" the command to the `lxd-user` command that will be
statically compiled so that it can connect to the running LXD daemon via
the unix socket to indicate the container is stopping, even after refreshing
the LXD snap to a different core base snap.

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>

simondeziel commented on snapcraft/commands/daemon.start:
```diff
@@ -652,8 +652,8 @@ if [ -n "${PID}" ] && [ "$(readlink "/proc/self/ns/mnt")" != "$(readlink "/proc/
-## Check for existing LXDs
-for pid in $(pgrep --euid 0 --full "lxd(\.debug)? --logfile"); do
+## Check for existing LXDs (check for bare lxd and lxd-daemon processes not running via /bin/sh, as well as lxd.debug processes).
+for pid in $(pgrep --euid 0 --full "(^lxd-daemon|^lxd|lxd.debug) --logfile"); do
```
Optional change:

```suggestion
for pid in $(pgrep --euid 0 --full "^lxd(|-daemon|\.debug) --logfile"); do
```
Which reads more naturally to me? Starts with `lxd` followed by nothing or `-daemon` or `.debug` suffix followed by `--logfile`.

tomponline replied:
the `^` bit must only be for lxd-daemon and lxd but not lxd.debug because of the way its invoked it will be prefixed with the path.

simondeziel replied:
Very true, thanks! Then the only missing bit is escaping the `.` of `.debug` (`\.debug`).

simondeziel reviewed (commented):
While I see no technical issue with the PR, I'd rather avoid us moving away from `lxd` as the exec name for our daemon.

Unfortunately, I don't have anything better to propose at the time :/

tomponline commented:
> While I see no technical issue with the PR, I'd rather avoid us moving away from `lxd` as the exec name for our daemon.

Me too. Ive got a (simple) idea, will try in the morning.

tomponline commented:
The solution was to put the lxd binary in `/sbin` and the stop hook wrapper script in `/bin`.

tomponline committed:
lxd: Build lxd server binary into /sbin

    The PATH inside the snap is configured to prefer /sbin over /bin so the lxd binary will still
    be called by default.

    The lxd-stophook-wrapper script which will be placed into `/bin/lxd`.

    Because containers are configured to call `/snap/lxd/current/bin/lxd callhook` which in turn will be
    routed to `/snap/lxd/current/bin/lxd-user` by the lxd-stophook-wrapper script, which will then connect
    to LXD's unix socket to indicate to the server that the container is stopping.

    Because lxd-user is compiled as a static binary this will work across snap core base changes.

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>

tomponline committed:
snapcraft: Use --strip-all for clarity

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>

simondeziel commented on lxd-stophook-wrapper/lxd:
```diff
@@ -0,0 +1,8 @@
+# This avoids polluting the process tree with this wrapper script.
+if [ "$1" = "callhook" ]; then
+    exec /snap/lxd/current/bin/lxd-user "$@"
+else
+    exec lxd "$@"
+fi
```
Just a question of style but elsewhere, we put the second `exec` outside of any `if/else` clause as the one inside the `if` is terminal anyway. https://github.com/canonical/lxd-pkg-snap/blob/latest-edge/snapcraft/wrappers/run-host#L10-L14 is one such example.

tomponline replied:
yeah lets do that

simondeziel commented on snapcraft/wrappers/lxd-stophook:
```diff
@@ -0,0 +1,8 @@
+# This avoids polluting the process tree with this wrapper script.
+if [ "$1" = "callhook" ]; then
+    exec /snap/lxd/current/bin/lxd-user "$@"
+else
```
`s/else/fi/` :)

tomponline replied:
I fixed this but must not have pushed it after hitting the issue with the snapd unconfined rule changes

simondeziel commented on snapcraft.yaml:
```diff
@@ -1602,6 +1612,7 @@ parts:
     organize:
       commands/snap-query: bin/
       hooks/: snap/hooks/
+      wrappers/lxd-stophook: bin/lxd
```
I missed that change in my previous review. I much prefer it to the previous extra part so thanks!

simondeziel commented on snapcraft/wrappers/lxd-stophook:
```diff
@@ -0,0 +1,8 @@
+    exec /snap/lxd/current/bin/lxd-user "$@"
+fi
+
+exec lxd "$@"
```
@tomponline It just occurred to me that we may not want to fallback to invoking `lxd` (and relying on `$PATH`) for the non-callhook scenario.

If we ended up in that `lxd-stophook` script for a non-callhook, I think it'd be cleaner to `exit 1` with an error message explaining what's wrong.

tomponline replied:
noted, thanks, will try this!
